File: 2019/kickstart/roundg/b.js
var readline = require('readline');

var rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});
var lines = [];
rl.on('line', function(input) {
    lines.push(input);
});
rl.on('close', function() {
    var t = parseInt(lines[0]);
    var l = 1;
    for (var i = 0; i < t; i++) {
        var tokens = lines[l++].split(' ').map((x) => +x);
        var arr = lines[l++].split(' ').map((x) => +x);
        console.log('Case #%d: %s', i + 1, solve(tokens[0], tokens[1], arr));
    }
});

function solve(n, m, arr) {
    var max = Math.max.apply(Math, arr);
    max = Math.max(max, Math.floor(m / n));
    var r = 1;
    while (max) {
        max >>= 1;
        r <<= 1;
    }
    r >>= 1;
    r = Math.max(r, 1);

    var ones = {}; // r -> c
    for (var i = r; i; i >>= 1) {
        ones[i] = arr.reduce((s, a) => s + ((a & i) ? 1 : 0), 0);
    }

    // search bit by bit, left to right
    var q = [{ r, k: 0, s: 0 }];
    var res = -1;
    while (q.length) {
        var state = q.shift();

        var valid = false;
        var oneBits = ones[state.r];
        // if (oneBits >= arr.length / 2) {
        if (!valid) {
            // use 1
            var sum = state.s + (arr.length - oneBits) * state.r;
            if (sum <= m) {
                if (state.r == 1) {
                    res = Math.max(res, state.k + state.r);
                    continue;
                }
                q.push({r: state.r / 2, k: state.k + state.r, s: sum});
                valid = true;
            }
        }
        // if (oneBits <= arr.length / 2) {
        if (!valid) {
            // use 0
            var sum = state.s + oneBits * state.r;
            if (sum <= m) {
                if (state.r == 1) {
                    res = Math.max(res, state.k);
                    continue;
                }
                q.push({r: state.r / 2, k: state.k, s: sum});
            } else {
                break;
            }
        }
    }

    return res;
}

File: 2019/kickstart/roundg/b.py
import math

def solve(n, m, arr):
    mx = max(arr)
    mx = max(mx, math.floor(m / n))

    r = 1
    while mx:
        mx = math.floor(mx / 2)
        r *= 2
    r /= 2
    r = max(r, 1)

    ones = {} # r -> c
    i = r
    while i >= 1:
        ones[i] = reduce(lambda s, a: s + (1 if (a & i) else 0), arr, 0)
        i = i / 2

    # precompute the min
    agg = 0
    mn = {} # r -> min
    i = 1
    while i <= r:
        agg = agg + i * min(ones[i], len(arr) - ones[i])
        mn[i] = agg
        i = i * 2

    # search bit by bit, left to right
    state = {"r": r, "k": 0, "s": 0}
    i = r
    res = -1
    while i >= 1:
        oneBits = ones[state["r"]]
        s = state["s"] + state["r"] * (len(arr) - oneBits)
        if s <= m and (state["r"] == 1 or s + mn[state["r"] / 2] <= m):
            if state["r"] == 1:
                res = max(res, state["k"] + state["r"])
            state = {"r": state["r"] / 2, "k": state["k"] + state["r"], "s": s}
        else:
            s = state["s"] + state["r"] * oneBits
            if s <= m:
                if state["r"] == 1:
                    res = max(res, state["k"])
                state = {"r": state["r"] / 2, "k": state["k"], "s": s}
            else:
                break
        i = i / 2

    return res

t = int(raw_input())
for i in xrange(1, t + 1):
    n, m = map(int, raw_input().split(' '))
    arr = map(int, raw_input().split(' '))

    print 'Case #{}: {}'.format(i, solve(n, m, arr))
